### PR TITLE
Implement the SKS reconciler for pod addressability

### PR DESCRIPTION
**Pull request by markusthoemmes**
Ref knative/serving#10751

This implements the actual reconciler and its resources. Note that the reconciler still isn't hooked into main so it won't be running after merging this.

### CODE REVIEW

File: pkg/reconciler/accessor/istio/destinationrule.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (vagababov) [inline pkg/reconciler/accessor/istio/destinationrule.go:2]: 2021

Comment (vagababov) [reply pkg/reconciler/accessor/istio/destinationrule.go:2]: Also why did linter ignore?

File: pkg/reconciler/accessor/istio/destinationrule.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	istioclientset "knative.dev/net-istio/pkg/client/istio/clientset/versioned"
+	istiolisters "knative.dev/net-istio/pkg/client/istio/listers/networking/v1alpha3"
+	kaccessor "knative.dev/net-istio/pkg/reconciler/accessor"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/kmeta"
+)
+
+// DestinationRuleAccessor is an interface for accessing DestinationRule.
+type DestinationRuleAccessor interface {
```
Comment (vagababov) [inline pkg/reconciler/accessor/istio/destinationrule.go:36]: does the type has to be pulic?

File: pkg/reconciler/accessor/istio/destinationrule.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+// ReconcileDestinationRule reconciles DestinationRule to the desired status.
+func ReconcileDestinationRule(ctx context.Context, owner kmeta.Accessor, desired *v1alpha3.DestinationRule,
+	drAccessor DestinationRuleAccessor) (*v1alpha3.DestinationRule, error) {
+
+	recorder := controller.GetEventRecorder(ctx)
+	if recorder == nil {
+		return nil, fmt.Errorf("recorder for reconciling DestinationRule %s/%s is not created", desired.Namespace, desired.Name)
+	}
+	ns := desired.Namespace
+	name := desired.Name
```
Comment (vagababov) [inline pkg/reconciler/accessor/istio/destinationrule.go:56]:
```suggestion
	ns, name := desired.Namespace, desired.Name
```

File: pkg/reconciler/accessor/istio/destinationrule_test.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+func (f *FakeDestinatioRuleAccessor) GetIstioClient() istioclientset.Interface {
+	return f.client
+}
+
+func (f *FakeDestinatioRuleAccessor) GetDestinationRuleLister() istiolisters.DestinationRuleLister {
+	return f.drLister
+}
+
+func TestReconcileDestinationRule_Create(t *testing.T) {
```
Comment (vagababov) [inline pkg/reconciler/accessor/istio/destinationrule_test.go:73]:
```suggestion
func TestReconcileDestinationRuleCreate(t *testing.T) {
```

File: pkg/reconciler/accessor/istio/destinationrule_test.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	})
+
+	ReconcileDestinationRule(ctx, ownerObj, desiredDR, accessor)
+
+	if err := h.WaitForHooks(3 * time.Second); err != nil {
+		t.Error("Failed to Reconcile DestinationRule:", err)
+	}
+}
+
+func TestReconcileDestinationRule_Update(t *testing.T) {
```
Comment (vagababov) [inline pkg/reconciler/accessor/istio/destinationrule_test.go:110]:
```suggestion
func TestReconcileDestinationRuleUpdate(t *testing.T) {
```

File: pkg/reconciler/accessor/istio/destinationrule_test.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+		if diff := cmp.Diff(got, desiredDR); diff != "" {
+			t.Log("Unexpected DestinationRule (-want, +got):", diff)
+			return HookIncomplete
+		}
+		return HookComplete
+	})
+
+	ReconcileDestinationRule(ctx, ownerObj, desiredDR, accessor)
+
+	if err := h.WaitForHooks(3 * time.Second); err != nil {
```
Comment (vagababov) [inline pkg/reconciler/accessor/istio/destinationrule_test.go:105]: Let's not do hooks (i've removed all in serving, I think), they're flaky. Do just wait/poll.

File: pkg/reconciler/accessor/istio/destinationrule.go (hunk 2 above)
Comment (markusthoemmes) [reply pkg/reconciler/accessor/istio/destinationrule.go]: Yep, see usage in this PR.

File: pkg/reconciler/accessor/istio/destinationrule.go (hunk 3 above)
Comment (markusthoemmes) [reply pkg/reconciler/accessor/istio/destinationrule.go]: Has been sent (and merged) as a prereq PR already.

File: pkg/reconciler/accessor/istio/destinationrule_test.go (hunk 4 above)
Comment (markusthoemmes) [reply pkg/reconciler/accessor/istio/destinationrule_test.go]: Has been sent (and merged) as a prereq PR already.

File: pkg/reconciler/accessor/istio/destinationrule_test.go (hunk 5 above)
Comment (markusthoemmes) [reply pkg/reconciler/accessor/istio/destinationrule_test.go]: Has been sent (and merged) as a prereq PR already.

File: pkg/reconciler/accessor/istio/destinationrule_test.go (hunk 6 above)
Comment (markusthoemmes) [reply pkg/reconciler/accessor/istio/destinationrule_test.go]: Has been sent (and merged) as a prereq PR already.

Feel free to clean the entire repo out though.

File: pkg/reconciler/serverlessservice/resources/destinationrule.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	subsetNormal = "normal"
+	subsetDirect = "direct"
+)
+
+// MakeDestinationRule creates a DestinationRule that defines a "normal" and a "direct"
+// loadbalancer for the service in question, to allow for pod addressability, even in mesh.
+func MakeDestinationRule(sks *v1alpha1.ServerlessService) *v1alpha3.DestinationRule {
+	ns := sks.Namespace
+	name := kmeta.ChildName(sks.Name, "-private")
+	host := fmt.Sprintf("%s.%s.svc.%s", name, ns, pkgnetwork.GetClusterDomainName())
```
Comment (julz) [inline pkg/reconciler/serverlessservice/resources/destinationrule.go:41]: might as well use [GetServiceHostname](https://github.com/knative/networking/blob/1b3d24e5be444c0938880826602f6e17da8dd279/vendor/knative.dev/pkg/network/domain.go#L39) helper in same package here

File: pkg/reconciler/serverlessservice/resources/virtualservice.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	"knative.dev/pkg/kmeta"
+	pkgnetwork "knative.dev/pkg/network"
+)
+
+// MakeVirtualService creates a placeholder virtual service to allow direct
+// pod addressability, even for mesh cases.
+func MakeVirtualService(sks *v1alpha1.ServerlessService) *v1alpha3.VirtualService {
+	ns := sks.Namespace
+	name := kmeta.ChildName(sks.Name, "-private")
+	host := fmt.Sprintf("%s.%s.svc.%s", name, ns, pkgnetwork.GetClusterDomainName())
```
Comment (julz) [inline pkg/reconciler/serverlessservice/resources/virtualservice.go:37]: same ^

File: pkg/reconciler/serverlessservice/serverlessservice_test.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+		r := &reconciler{
+			istioclient:           istioclient.Get(ctx),
+			virtualServiceLister:  listers.GetVirtualServiceLister(),
+			destinationRuleLister: listers.GetDestinationRuleLister(),
+		}
+
+		return sksreconciler.NewReconciler(ctx, logging.FromContext(ctx), fakenetworkingclient.Get(ctx),
+			listers.GetServerlessServiceLister(), controller.GetEventRecorder(ctx), r, controller.Options{
+				ConfigStore: &testConfigStore{
+					// Enable reconciling gateway.
```
Comment (julz) [inline pkg/reconciler/serverlessservice/serverlessservice_test.go:194]: Probably missing something obvious, but not sure what this comment refers to. We reconcile SKS to VS and DR, so no gateway really involved?

Comment (markusthoemmes) [reply pkg/reconciler/serverlessservice/serverlessservice_test.go:194]: copy aaaaand pasta :D

File: pkg/reconciler/serverlessservice/resources/destinationrule.go (hunk 7 above)
Comment (markusthoemmes) [reply pkg/reconciler/serverlessservice/resources/destinationrule.go]: Cool didn't know that existed. Thanks!

**Review by julz — APPROVED** [review]

**Review by vagababov — APPROVED** [review]
/lgtm

**Pull request merged**